### ISSUE TITLE
one-var: allow both `const` and `let`

### ISSUE THREAD

**Issue by joeybaker** [eslint/eslint#2301]
Currently, with `"one-var": 1`, the following code with cause a warning:

```js
function(){
  let a = true;
  const b = false;
}
```

`"one-var"` wants all variable declarations to be declared together, so this causes a warning. However, it's legitimate code, and might be intended. 

I suggest, via PR #2300, that we add an additional value to the first option `"var-only"`. This will cause:

The following patterns are considered warnings:

```js
function foo() {
    var bar;
    var baz;
}

function foo() {
    var bar;

    if (baz) {
        var qux = true;
    }
}
```

The following patterns are not considered warnings:

```js
function foo() {
    var bar,
        baz;
}

function foo() {
    var bar,
        qux;

    if (baz) {
        qux = true;
    }
}

function foo() {
    const baz = true;
    let bar;

    if (baz) {
        bar = true;
    }
}
```

Adding the option is backward compatible.

**Comment by nzakas** [comment]
Interesting case. I think this opens up a larger discussion around this rule and ES6 in general. 

You could argue that this rule should apply to each *scope*, so that there must be one `var` statement per function and also one `let` statement and one `const` statement per block. With that argument, there wouldn't be the need for a new option, it would just be the default behavior.

Unless someone argues that you'd want one `let` statement per function...though I'd have a hard time buying that argument without some convincing.

**Comment by joeybaker** [comment]
Right, I doubt that anyone wants one `let` per block. I also didn't want to change default behavior because that would be a breaking change.

I'd kinda like `one-var` to deal only with `var` and let `let` and `const` be managed by their own rule if necessary, but, that is a pretty big breaking change.

**Comment by nzakas** [comment]
Don't presume to know what everyone wants or doesn't want. :)

Most of our rules related to `var` also apply to `let` and `const`, so I don't think it makes sense to duplicate this one for that purpose. 

I'm also not adverse to breaking changes if it's the right thing to do. Correctness trumps backwards compatibility, and it's unclear right now if the way the rule works right now is a bug or not.

**Comment by joeybaker** [comment]
Heh… defaults are assumptions. That's all I mean :) 

If you're not adverse to breaking changes what do you recommend?

**Comment by nzakas** [comment]
I'm not entirely sure, I'm trying to start a discussion here. Some options would be to always ignore let and const, or only allow one per block.

**Comment by joeybaker** [comment]
What do you think of changing `one-var` to only enforce `var` and adding rules for `one-let` and `one-const`?

**Comment by nzakas** [comment]
Not a fan of having more than one rule doing the exact same thing, it becomes a maintenance issue. If we want `let` and `const` to follow the same rule as `var`, then it makes sense to keep it a single rule.

**Comment by joeybaker** [comment]
Yea, I agree. That's why I landed on adding the new value for the option `"one-var"`. That allows us to keep things at one rule, and gives people the flexibility to enforce the rule for `var`, but disable it for `const` and `let` so that you can use all three in a function and not have a false error.

Though, I've since realized: var is scoped to a function and `let` and `const` are scoped to a block, meaning that rule will still report false positives while using `let` and `const`. I'm actually coming to the opinion that this should only apply to `var` and ignore `let` and `const`.

**Comment by nzakas** [comment]
@ilyavolodin @btmills @lo1tuma any opinions on this?

**Comment by ilyavolodin** [comment]
If we don't want to create `one-let` and `one-const` rules, and have this rule handle all of them, then I think having options to control each and every one of them is a must. I personally don't currently use ES6, so I don't have a strong opinion on style, however, from other languages, it's very common to want all variables to be declared on one line, but have constants declared separately from each other. Const names are usually more verbose and uppercased and having a long line full of screaming text is not a good idea.
 Like in C# I would do the following:
```c#
private int a = 0, b = 0, c = 0;
private const int MY_CONSTANT = 22;
private const int MY_OTHER_CONST = 33;
```
If you would choose to go that route, then I would suggest not doing option `var-only`, I would go with the following construction of options:
```
one-var: [2, [{vars: true}, {let: true, scope: function}, {const: false}]]
```

**Comment by btmills** [comment]
I agree with previous comments that (a) it's best to have one rule handling all three types, but (b) it should be possible, though not required, to configure each separately.

However, I think it's a bad idea to expand the scope of `one-var` to include specifying block or function scope for block bindings. That seems to me better handled by a hypothetical `no-block-scope` rule.

For configuration, my proposal is similar to @ilyavolodin's, minus options for scope:

```js
{
    // (default) Exactly one variable declaration per type per function (var) or block (let or const)
    "one-var": [2, "always"],

    // Exactly one declarator per declaration per function (var) or block (let or const)
    "one-var": [2, "never"],

    // Configure each declaration type individually. Defaults to "always" if key not present.
    "one-var": [2, {
        // "var": "always", // Exactly one var declaration per function
        "let": "always", // Exactly one let declaration per block
        "const", "never" // Exactly one declarator per const declaration per block
    }]
}
```

**Comment by ilyavolodin** [comment]
@btmills The reason for including scope was the following code that I use on the daily bases:
```
let some = [];

for (let i=0, l=some.length; i<l; i++) {
...
}
```

Without scope that would error.
Other then that, your options look better then what I suggested.

**Comment by nzakas** [comment]
@ilyavolodin he's proposing that let and const be evaluated per block, so your example would pass.

**Comment by joeybaker** [comment]
:+1: @btmills API from me.

**Comment by joeybaker** [comment]
@nzakas would you like a PR to implement @btmills' API?

**Comment by nzakas** [comment]
Yup, looks good.
